release-23.2: server/license: Change how trial license usage is tracked

spilchen opened the pull request:
Backport 1/1 commits from #132177.

/cc @cockroachdb/release

---

There’s a race condition when updating the enterprise.license config setting and checking the trial usage count. If a node starts up while a new trial license is being applied, it can encounter an issue where it sees the updated trial usage count in KV before receiving the corresponding enterprise.license config setting. This causes the license update to be rejected, as it incorrectly assumes a trial license has already been used.

This change addresses the issue by modifying what is stored in the KV for the trial license. Instead of tracking the number of trial licenses used, which would ever be 0 or 1, we now store the expiry timestamp of any active or past trial license. The enterprise.license validation function will compare the expiry of the new license against the cached value from KV. If the expiry timestamp is not set or matches the expiry of the new license, the validation will proceed. Otherwise, it will fail as before.

This change will be backported to 24.2, 24.1, 23.2 and 23.1.

Epic: CRDB-39988
Closes #131968
Release note: none
Release justification: This work is part of the CockroachDB core deprecation.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/132228)
<!-- Reviewable:end -->

spilchen committed:
server/license: Change how trial license usage is tracked

There’s a race condition when updating the enterprise.license config setting
and checking the trial usage count. If a node starts up while a new trial
license is being applied, it can encounter an issue where it sees the updated
trial usage count in KV before receiving the corresponding enterprise.license
config setting. This causes the license update to be rejected, as it
incorrectly assumes a trial license is already in use.

This change addresses the issue by modifying what is stored in the KV for the
trial license. Instead of tracking the number of installed trial licenses, we
now store the expiry timestamp of any active or past trial license. The
enterprise.license validation function will compare the expiry of the new
license against the cached value from KV. If the expiry timestamp is not set or
matches the expiry of the new license, the validation will proceed. Otherwise,
it will fail as before.

Epic: CRDB-39988
Closes #131968
Release note: none
